diff --git a/wrapper/target.cpp b/wrapper/target.cpp
index 82bf65c40..a81ce97b7 100644
--- a/wrapper/target.cpp
+++ b/wrapper/target.cpp
@@ -741,6 +741,10 @@ bool Target::setup() {
           (stdlib == StdLib::libstdcxx && usegcclibs)) {
         fargs.push_back("-nostdinc++");
         fargs.push_back("-Qunused-arguments");
+
+        if ((SDKOSNum >= OSVersion(11, 1)) && (stdlib == StdLib::libcxx)) {
+          fargs.push_back("-lc++");
+        }
       }

       if (stdlib == StdLib::libstdcxx && usegcclibs && targetarch.size() < 2 &&
